### PR TITLE
[SNAP-2065] S2 msi reader with gdal

### PR DESCRIPTION
**Pull request by oanahogoiu**

### CODE REVIEW

File: lib-gdal-preferences-ui/pom.xml [hunk 1]
```diff
@@ -33,6 +33,7 @@
         <dependency>
             <groupId>org.esa.snap</groupId>
             <artifactId>lib-gdal</artifactId>
+            <version>${project.parent.version}</version>
```
Comment (marpet) [inline lib-gdal-preferences-ui/pom.xml:36]: better replace by ${snap.version} to be consistent with usage in other modules.

File: lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanelController.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import java.beans.PropertyChangeListener;
+import java.beans.PropertyChangeSupport;
+
+/**
+ * GDAL Options Panel Controller for GDAL native library loader.
+ * Used for provide an controller for UI to the strategy with loading GDAL native library.
+ *
+ * @author Adrian Draghici
+ */
+@OptionsPanelController.SubRegistration(
```
Comment (marpet) [inline lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanelController.java:17]: Can we remove the references to S2TBX here?
This is now a general SNAP GFAL Loader

File: lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanel.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
+package org.esa.s2tbx.dataio.gdal.preferences;
```
Comment (marpet) [inline lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanel.java:1]: we should replace org.esa.s2tbx with org.esa.snap

**Review by marpet — CHANGES_REQUESTED** [review]
In general okay. But the references to s2tbx should be removed. See the comments.

File: lib-gdal-preferences-ui/pom.xml (hunk 1 above)
Comment (oanahogoiu) [reply lib-gdal-preferences-ui/pom.xml]: Updates made.

File: lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanelController.java (hunk 2 above)
Comment (oanahogoiu) [reply lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanelController.java]: Updates made.

File: lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanel.java (hunk 3 above)
Comment (oanahogoiu) [reply lib-gdal-preferences-ui/src/main/java/org/esa/s2tbx/dataio/gdal/preferences/GDALOptionsPanel.java]: Updates made.

**Comment by marpet** [comment]
Thanks for the updated. One more thing ...
Is there a related JIRA ticket for this work? It should be mentioned in the PR title. If not, the ticket should be created.

**Comment by oanahogoiu** [comment]
JIRA ticket [SIITBX-501].

**Comment by marpet** [comment]
I just checked out the two branches and rebuilt everything. I still get the error:

```
java.io.EOFException
	at javax.imageio.stream.ImageInputStreamImpl.readShort(ImageInputStreamImpl.java:229)
	at javax.imageio.stream.ImageInputStreamImpl.readUnsignedShort(ImageInputStreamImpl.java:242)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFIFD.initialize(TIFFIFD.java:241)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageMetadata.initializeFromStream(TIFFImageMetadata.java:148)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.readMetadata(TIFFImageReader.java:911)
Caused: javax.imageio.IIOException: I/O error reading image metadata!
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.readMetadata(TIFFImageReader.java:917)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.seekToImage(TIFFImageReader.java:864)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.seekToImage(TIFFImageReader.java:822)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.getWidth(TIFFImageReader.java:1080)
	at org.esa.snap.jp2.reader.internal.JP2TileOpImage$ImageReader.getImageWidth(JP2TileOpImage.java:306)
	at org.esa.snap.jp2.reader.internal.JP2TileOpImage.computeRectIndirect(JP2TileOpImage.java:148)
	at org.esa.snap.jp2.reader.internal.JP2TileOpImage.computeRect(JP2TileOpImage.java:113)
Caused: java.lang.IllegalStateException: Failed to read the data for level 0 and rectangle java.awt.Rectangle[x=0,y=0,width=256,height=256].
	at org.esa.snap.jp2.reader.internal.JP2TileOpImage.computeRect(JP2TileOpImage.java:115)
	at javax.media.jai.SourcelessOpImage.computeTile(Unknown Source)
```

**Comment by oanahogoiu** [comment]
Please clear all SNAP cache before testing on these branches:
C:\Users\[username]\.snap
C:\Users\[username]\AppData\Roaming\SNAP or C:\Users\[username]\AppData\Local\SNAP
C:\Users\[username]\AppData\Local\Temp\snap-[username]

JP2 reader is not used anymore, that was the purpose, to replace it with GDAL.

**Comment by marpet** [comment]
Now I'm getting this exception when trying to open any S2 product:

```
java.io.IOException: Cannot open E:\SENTINEL2\S2B_MSIL1C_20200620T160829_N0209_R140_T16QCF_20200620T193303.SAFE\GRANULE\L1C_T16QCF_A017183_20200620T162352\IMG_DATA\T16QCF_20200620T160829_B01.jp2
	at org.esa.s2tbx.dataio.s2.ortho.Sentinel2OrthoProductReader.addBands(Sentinel2OrthoProductReader.java:594)
	at org.esa.s2tbx.dataio.s2.ortho.Sentinel2OrthoProductReader.readProduct(Sentinel2OrthoProductReader.java:268)
	at org.esa.s2tbx.dataio.s2.Sentinel2ProductReader.readProductNodesImpl(Sentinel2ProductReader.java:198)
	at org.esa.snap.core.dataio.AbstractProductReader.readProductNodes(AbstractProductReader.java:178)
	at org.esa.s2tbx.dataio.s2.ortho.Sentinel2OrthoProductReaderProxy.readProductNodes(Sentinel2OrthoProductReaderProxy.java:90)
	at org.esa.snap.core.dataio.ProductIO.readProduct(ProductIO.java:182)
```
This is the full log:
[log.txt](https://github.com/senbox-org/snap-desktop/files/10899465/log.txt)

**Comment by oanahogoiu** [comment]
From the log, the cause seems to be:
ERROR 1: PROJ: proj_create_from_database: Cannot find proj.db

Please open a cmd, type "echo %PROJ_LIB%" and tell us the result.

**Comment by marpet** [comment]
It is just %PROJ_LIB%

image

**Comment by a-draghici** [comment]
> It is just %PROJ_LIB%
> 
> image
It confirms that the environment variable `PROJ_LIB` is not set on the system.
Please tell us the result of execution for `where gdalinfo`.
Thank you.

**Comment by marpet** [comment]
image
Responds translated:
INFORMATION: No files with the specified pattern could be found.

**Comment by a-draghici** [comment]
> image Responds translated: INFORMATION: No files with the specified pattern could be found.

It confirms that the GDAL is not installed on the system.
Please send us the `gdal_dir_tree.txt` file with result of execution for `tree /f /a %USERPROFILE%\.snap\auxdata\gdal > %USERPROFILE%\gdal_dir_tree.txt`.
Thank you.

**Comment by marpet** [comment]
[gdal_dir_tree.txt](https://github.com/senbox-org/snap-desktop/files/10908137/gdal_dir_tree.txt)

In the log it is said:
```
INFO [org.esa.snap.dataio.gdal.GDALVersion]: GDAL not found on system. Internal GDAL 3.2.1 from distribution will be used. (f0)
INFO [org.esa.snap.dataio.gdal.GDALVersion]: Internal GDAL 3.2.1 set to be used by SNAP.
```

**Comment by oanahogoiu** [comment]
Marco, do you have the same issue on another branches too? (e.g. 9.x or master) Ot it's just on this PR branch? 
I am asking this because none of us managed to reproduce it until now and the way of loading GDAL libraries was not changed on this branch.

**Comment by marpet** [comment]
I haven't checked with other branches again. I'll try and let you know

**Comment by marpet** [comment]
Switching to the master branch and rebuilding the project, reading S2 products still works. Also, from the installed SNAP it works.
Switching back to the three branches related to this and rebuilding everything it is not working anymore.
image

**Comment by a-draghici** [comment]
Marco, please test again. Not reproduced but I tried a potential solution. You can also activate FINE logging level, maybe more details will be shown.

**Review by marpet — APPROVED** [review]
Now it is working!!!
I also tried to resample a product and it worked. 
So it seems this was the last missing piece.

**Pull request merged**